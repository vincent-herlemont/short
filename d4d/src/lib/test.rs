//! Helper for test related of d4d domain.
use crate::lib::resource;
use crate::lib::resource::Resource;
use std::fs;
use std::path::PathBuf;
use tempdir::TempDir;

pub struct Config {
    pub tmp_dir: PathBuf,
}

impl Drop for Config {
    fn drop(&mut self) {
        fs::remove_dir_all(self.tmp_dir.clone()).expect("can not clean tmp directory");
    }
}

/// Return [`InspectorConfig`], create temporary directory and copy resource on it.
///
/// The temporary directory is owned by [`InspectorConfig.path`].
///
/// # Recommendation
///
/// Need to call [`after`] at the end of test.
pub fn before(test_name: &str) -> Config {
    let test_name = format!("{}.{}", "d4d", test_name);

    // Create temporary directory.
    let path = TempDir::new(test_name.as_str())
        .expect("fail to create temporary directory")
        .into_path();

    // Copy resources to it.
    resource::to_dir(&path).expect("fail to copy resources");

    Config { tmp_dir: path }
}

/// Assert that patter value or|and expression is present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_find!(v, 2);
///assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
///assert_find!(v, el, el < &&5)
///assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}
///
/// ```
#[macro_export]
macro_rules! assert_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, true)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, true)
    };
    ($v:ident,$arms:pat,$e:expr,$b:expr) => {
        if (&$v)
            .iter()
            .find(|el| match el {
                $arms => $e,
                _ => false,
            })
            .is_some()
            != $b
        {
            panic!(format!(
                "assertion failed: {} {{{}}} {} in {{{}}}",
                if ($b) {
                    String::from("can not found")
                } else {
                    String::from("found")
                },
                String::from(stringify!($arms)),
                if (stringify!($e) == "true" || stringify!($e) == "false") {
                    String::from("")
                } else {
                    String::from(format!(
                        "with expresion {{{}}}",
                        String::from(stringify!($e))
                    ))
                },
                String::from(stringify!($v))
            ));
        };
    };
}

/// Assert that patter value or|and expression is not present on an vector.
///
/// # Notice
/// The macro don't take the ownership of a vector.
///
/// You have to add [`#[allow(unreachable_patterns)]`] to avoid warning
///
/// # Example
/// ```
/// let v = vec![1,2,3,4];
///assert_not_find!(v, 8);
///assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
///assert_not_find!(v, el, el > &&5);
///assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
/// ```
#[macro_export]
macro_rules! assert_not_find {
    ($v:ident,$arms:pat) => {
        assert_find!($v, $arms, true, false)
    };
    ($v:ident,$arms:pat,$e:expr) => {
        assert_find!($v, $arms, $e, false)
    };
}

#[cfg(test)]
mod tests {

    #[allow(unreachable_patterns)]
    #[test]
    fn assert_find_macro_test() {
        let v = vec![1, 2, 3, 4];
        assert_find!(v, 2);
        // assert_find!(v, 8); // assertion failed: can not found {8}  in {v}
        assert_find!(v, el, el < &&5);
        // assert_find!(v, el, el > &&5); // assertion failed: can not found {el} with expresion {el > &&5} in {v}

        assert_not_find!(v, 8);
        // assert_not_find!(v, 2); // assertion failed: found {2}  in {v}
        assert_not_find!(v, el, el > &&5);
        //assert_not_find!(v, el, el < &&5); // assertion failed: found {el} with expresion {el < &&5} in {v}
    }
}

pub const TEST_RESOURCE_DIRECTORY: &'static str = "./init_tpl";

/// Get all [`Resource`]
pub fn get_resource() -> Vec<Resource> {
    vec![
        Resource::new(
            "./init_tpl/other_conf.yaml",
            include_str!("./init_tpl/0_other_conf.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate.yaml",
            include_str!("./init_tpl/1_certificate.yaml"),
        ),
        Resource::new(
            "./init_tpl/1_certificate_altered.yaml",
            include_str!("./init_tpl/2_certificate_altered.yaml"),
        ),
        Resource::new(
            "./init_tpl/3_test/0_test.js",
            include_str!("./init_tpl/3_test/0_test.js"),
        ),
        Resource::new(
            "./init_tpl/4_tpl_certificate/certificate.yaml",
            include_str!("./init_tpl/4_tpl_certificate/certificate.yaml"),
        ),
    ]
}
