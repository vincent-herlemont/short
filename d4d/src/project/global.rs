use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File, OpenOptions};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CURRENT_FILE_NAME: &'static str = "projects.yaml";

fn global_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_global_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let path = global_directory_path(home);
    match create_dir(path) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn global_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    global_directory_path(home).join(PROJECT_CURRENT_FILE_NAME)
}

#[allow(dead_code)]
fn read_global_file<P: AsRef<Path>>(home: P) -> Result<GlobalProjects> {
    let path = global_file_path(home);
    let file = File::open(path)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

/// Create or overwrite project config file.
#[allow(dead_code)]
fn save_global_file<P: AsRef<Path>>(home: P, projects: &GlobalProjects) -> Result<()> {
    let path = global_file_path(home);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &projects)?;
    Ok(())
}

#[derive(Serialize, Deserialize, Debug)]
struct GlobalProject {
    name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    path: Option<PathBuf>,

    #[serde(skip_serializing_if = "Option::is_none")]
    current_env: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    git_secret_repo: Option<String>,
}

impl GlobalProject {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> GlobalProject {
        GlobalProject {
            name: String::from(name.as_ref()),
            path: None,
            current_env: None,
            git_secret_repo: None,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GlobalProjects {
    #[serde(skip)]
    home_dir: PathBuf,

    #[serde(rename = "projects")]
    all: Vec<GlobalProject>,
}

impl GlobalProjects {
    pub fn new<P: AsRef<Path>>(home_dir: P) -> Result<GlobalProjects> {
        let home_dir = home_dir.as_ref().to_path_buf();
        match read_global_file(&home_dir) {
            Ok(global_file) => Ok(global_file),
            Err(_) => {
                let global_directory = global_directory_path(&home_dir);
                if !global_directory.exists() {
                    match create_global_directory(&home_dir) {
                        Ok(_) => (),
                        Err(err) => {
                            return Err(Error::wrap(
                                format!(
                                    "fail to create configuration directory {}",
                                    global_directory.to_string_lossy()
                                ),
                                err,
                            ));
                        }
                    }
                }
                // TODO: match for create err only if file does not exist.
                let global_projects = GlobalProjects {
                    home_dir: home_dir.to_owned(),
                    all: vec![],
                };
                match save_global_file(&home_dir, &global_projects) {
                    Ok(_) => Ok(global_projects),
                    Err(err) => Err(Error::wrap(
                        format!(
                            "fail to create global file to {}",
                            home_dir.to_string_lossy()
                        ),
                        err,
                    )),
                }
            }
        }
    }

    pub fn add<N, P>(&mut self, name: N, path: P) -> Result<()>
    where
        N: AsRef<str>,
        P: AsRef<Path>,
    {
        let path = PathBuf::from(path.as_ref())
            .canonicalize()
            .map_err(|e| Error::wrap("fail to get absolute path of : {}", Error::from(e)))?;

        self.all.push(GlobalProject {
            name: String::from(name.as_ref()),
            path: Some(path),
            current_env: None,
            git_secret_repo: None,
        });

        if let Err(err) = save_global_file(&self.home_dir, self) {
            Err(Error::wrap(
                format!(
                    "fail to save global file to : {}",
                    self.home_dir.to_string_lossy(),
                ),
                err,
            ))
        } else {
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::project::global::{
        create_global_directory, global_file_path, read_global_file, save_global_file,
        GlobalProject, GlobalProjects,
    };
    use insta::assert_debug_snapshot;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use std::path::PathBuf;
    use utils::asset::Assets;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_save_global_file() {
        let config = before("test_save_global_file", Assets::All(HashMap::new()));
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![GlobalProject::new("project_1")],
        };
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects:\n  - name: project_1"));

        // Overwrite
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![],
        };
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects: []"));
    }

    #[test]
    fn test_empty_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_empty_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects:
    - name: project_1
    - name: project_2
      current_env: dev
    - name: project_3
      path: /todo/plop/
      current_env: dev
    - name: project_4
      current_env: prod
      git_secret_repo: "git@privategit.com" 
"#,
        );
        let config = before("test_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_global_directory() {
        let config = before("create_global_directory", Assets::All(HashMap::new()));
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_eq!(r.iter().count(), 2);
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_err());
    }
}
