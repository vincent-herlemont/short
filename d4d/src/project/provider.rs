use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};


use utils::error::Error;
use utils::result::Result;

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "name")]
pub enum ProviderCfg {
    #[serde(rename = "aws")]
    ConfAws(AwsCfg),

    #[serde(rename = "unknown")]
    None,
}

impl ProviderCfg {
    pub fn aws(&self) -> Result<&AwsCfg> {
        if let ProviderCfg::ConfAws(conf_aws) = self {
            Ok(conf_aws)
        } else {
            Err(Error::from("aws provider is not defined"))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AwsCfg {
    region: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    template_path: Option<PathBuf>,
}

impl AwsCfg {
    pub fn new<N: AsRef<str>>(region: N) -> Self {
        Self {
            region: region.as_ref().to_string(),
            template_path: None,
        }
    }

    pub fn set_template_path<P: AsRef<Path>>(&mut self, template_path: P) -> &Self {
        self.template_path = Some(template_path.as_ref().to_path_buf());
        self
    }

    pub fn template_path(&self) -> Result<PathBuf> {
        self.template_path.clone().ok_or(Error::from(format!(
            "template_path missing", // TODO: set template name
        )))
    }
}
