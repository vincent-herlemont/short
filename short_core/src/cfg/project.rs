use std::path::PathBuf;

pub trait ProjectCfg {
    fn path(&self) -> PathBuf;
}
