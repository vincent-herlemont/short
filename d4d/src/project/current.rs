use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File, OpenOptions};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CURRENT_FILE_NAME: &'static str = "current.yaml";

fn project_config_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_project_config_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let project_config_directory = project_config_directory_path(home);
    match create_dir(project_config_directory) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn project_current_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    project_config_directory_path(home).join(PROJECT_CURRENT_FILE_NAME)
}

#[allow(dead_code)]
fn load_project_current_file<P: AsRef<Path>>(home: P) -> Result<Projects> {
    let path = project_current_file_path(home);
    let file = File::open(path)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

/// Create or overwrite project config file.
#[allow(dead_code)]
fn save_project_current_file<P: AsRef<Path>>(home: P, projects: Projects) -> Result<()> {
    let path = project_current_file_path(home);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &projects)?;
    Ok(())
}

#[derive(Serialize, Deserialize, Debug)]
struct Project {
    name: String,

    #[serde(skip_serializing)]
    current_env: Option<String>,

    #[serde(skip_serializing)]
    git_secret_repo: Option<String>,
}

impl Project {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> Project {
        Project {
            name: String::from(name.as_ref()),
            current_env: None,
            git_secret_repo: None,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Projects {
    #[serde(rename = "projects")]
    all: Vec<Project>,
}

#[cfg(test)]
mod tests {
    use crate::project::current::{
        create_project_config_directory, load_project_current_file, project_current_file_path,
        save_project_current_file, Project, Projects,
    };
    use insta::assert_debug_snapshot;
    use serde_yaml;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use utils::asset::Assets;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_save_project_current_file() {
        let config = before(
            "test_save_project_current_file",
            Assets::All(HashMap::new()),
        );
        let projects = Projects {
            all: vec![Project::new("project_1")],
        };
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r = save_project_current_file(&config.tmp_dir, projects);
        assert!(r.is_ok());
        let r = read_to_string(project_current_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects:\n  - name: project_1"));
        let read_create_projects_file = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(read_create_projects_file);
        let projects = Projects { all: vec![] };
        let r = save_project_current_file(&config.tmp_dir, projects);
        assert!(r.is_ok());
        let r = read_to_string(project_current_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects: []"));
        let read_overwrite_projects_file = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(read_overwrite_projects_file);
    }

    #[test]
    fn test_load_project_current_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/current.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_load_project_current_file", Assets::All(assets));
        let projects = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_project_config_directory() {
        let config = before(
            "create_project_config_directory",
            Assets::All(HashMap::new()),
        );
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_debug_snapshot!(r.iter().count());
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_err());
    }

    #[test]
    fn test_serialize_projects() {
        let root = Projects { all: vec![] };
        let s = serde_yaml::to_string(&root).unwrap();
        assert_debug_snapshot!(&s);
    }
}
