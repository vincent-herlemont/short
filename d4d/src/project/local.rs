use serde::export::Formatter;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Display;
use std::fs::OpenOptions;
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

const PROJECT_FILE_NAME: &'static str = "d4d.yaml";

fn local_file_path<P: AsRef<Path>>(root: P) -> PathBuf {
    root.as_ref().join(PROJECT_FILE_NAME)
}

fn save_local_file<P: AsRef<Path>>(root: P, local_projects: &LocalProjects) -> Result<()> {
    let path = local_file_path(root);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &local_projects).map_err(|e| Error::from(e))
}

fn read_local_file<P: AsRef<Path>>(root: P) -> Result<LocalProjects> {
    let file = OpenOptions::new().read(true).open(root)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

#[derive(Serialize, Deserialize, Debug)]
struct LocalProject {
    name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    template_path: Option<PathBuf>,

    #[serde(skip_serializing_if = "Option::is_none")]
    public_env_directory: Option<PathBuf>,
}

impl LocalProject {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> LocalProject {
        LocalProject {
            name: String::from(name.as_ref()),
            template_path: None,
            public_env_directory: None,
        }
    }
}

impl Display for LocalProject {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LocalProjects {
    #[serde(skip)]
    current_dir: PathBuf,

    #[serde(rename = "projects")]
    all: Vec<LocalProject>,
}

impl Display for LocalProjects {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        for local_project in &self.all {
            writeln!(f, " - {}", local_project)?;
        }
        Ok(())
    }
}

impl LocalProjects {
    pub fn new<P: AsRef<Path>>(current_dir: P) -> Result<LocalProjects> {
        let current_dir = current_dir.as_ref().to_path_buf();
        match read_local_file(&current_dir) {
            Ok(local_projects) => Ok(local_projects),
            Err(_) => {
                // TODO: match for create err only if file does not exist.
                let local_projects = LocalProjects {
                    current_dir: current_dir.to_owned(),
                    all: vec![],
                };
                match save_local_file(&current_dir, &local_projects) {
                    Ok(_) => Ok(local_projects),
                    Err(err) => Err(Error::wrap(
                        format!(
                            "fail to create local file {}",
                            current_dir.to_string_lossy()
                        ),
                        err,
                    )),
                }
            }
        }
    }

    pub fn add<N, TP, PED>(
        &mut self,
        name: N,
        template_path: TP,
        public_env_directory: PED,
    ) -> Result<()>
    where
        N: AsRef<str>,
        TP: AsRef<Path>,
        PED: AsRef<Path>,
    {
        self.all.push(LocalProject {
            name: String::from(name.as_ref()),
            template_path: Some(PathBuf::from(template_path.as_ref())),
            public_env_directory: Some(PathBuf::from(public_env_directory.as_ref())),
        });

        if let Err(err) = save_local_file(&self.current_dir, self) {
            Err(Error::wrap(
                format!(
                    "fail to save local file : {}",
                    self.current_dir.to_string_lossy()
                ),
                err,
            ))
        } else {
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::project::local::{
        local_file_path, read_local_file, save_local_file, LocalProject, LocalProjects,
    };
    use insta::assert_debug_snapshot;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use std::path::PathBuf;
    use utils::asset::Assets;
    use utils::test::before;

    #[test]
    fn test_read_local_file() {
        let mut assets = HashMap::new();
        assets.insert(
            "d4d.yaml",
            r#"
projects:
    - name: test_1
    - name: test_2
      template_path: "./test_template.yaml"
        "#,
        );
        let config = before("test_save_local_file", Assets::Static(assets));
        let local_projects = read_local_file(local_file_path(&config.tmp_dir)).unwrap();
        assert_debug_snapshot!(local_projects);
    }

    #[test]
    fn test_save_local_file() {
        let config = before("test_save_local_file", Assets::Static(HashMap::new()));
        let local_projects = LocalProjects {
            current_dir: PathBuf::new(),
            all: vec![LocalProject::new("test_1")],
        };
        let r = save_local_file(&config.tmp_dir, &local_projects);
        assert!(r.is_ok());
        let content = read_to_string(local_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(content, String::from("---\nprojects:\n  - name: test_1"));

        // Overwrite
        let local_projects = LocalProjects {
            current_dir: PathBuf::new(),
            all: vec![],
        };
        let r = save_local_file(&config.tmp_dir, &local_projects);
        assert!(r.is_ok());
        let content = read_to_string(local_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(content, String::from("---\nprojects: []"));
    }
}
