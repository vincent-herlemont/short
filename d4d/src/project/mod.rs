use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File};
use std::io::BufReader;
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CONFIG_FILE_NAME: &'static str = "projects.yaml";

fn project_config_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_project_config_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let project_config_directory = project_config_directory_path(home);
    match create_dir(project_config_directory) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn project_config_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    project_config_directory_path(home).join(PROJECT_CONFIG_FILE_NAME)
}

#[allow(dead_code)]
fn load_project_config_file<P: AsRef<Path>>(home: P) -> Result<Projects> {
    let project_config = project_config_file_path(home);
    let file = File::open(project_config)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

#[derive(Serialize, Deserialize, Debug)]
struct Project {
    name: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Projects {
    #[serde(rename = "projects")]
    all: Vec<Project>,
}

#[cfg(test)]
mod tests {
    use crate::project::{create_project_config_directory, load_project_config_file, Projects};
    use insta::assert_debug_snapshot;
    use serde_yaml;
    use std::collections::HashMap;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_load_project_config_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            "assets/.d4d/projects.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_load_project_config_directory", &assets);
        let projects = load_project_config_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_project_config_directory() {
        let config = before("create_project_config_directory", &HashMap::new());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_debug_snapshot!(r.iter().count());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
    }

    #[test]
    fn test_serialize_projects() {
        let root = Projects { all: vec![] };
        let s = serde_yaml::to_string(&root).unwrap();
        assert_debug_snapshot!(&s);
    }
}
