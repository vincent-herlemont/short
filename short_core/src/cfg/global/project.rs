use std::cell::RefCell;
use std::path::PathBuf;
use std::rc::Rc;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::cfg::global::setup::GlobalProjectSetupCfg;
use crate::cfg::{ProjectCfg, SetupCfg, SetupsCfg};

#[derive(Debug, Serialize, Deserialize)]
pub struct GlobalProjectCfg {
    file: PathBuf,
    setups: Rc<RefCell<Vec<Rc<RefCell<GlobalProjectSetupCfg>>>>>,
}

impl GlobalProjectCfg {
    pub fn new(file: &PathBuf) -> Result<Self> {
        let mut gp = GlobalProjectCfg {
            file: PathBuf::new(),
            setups: Rc::new(RefCell::new(vec![])),
        };
        gp.set_file(file)?;
        Ok(gp)
    }

    pub fn set_file(&mut self, file: &PathBuf) -> Result<()> {
        if (!file.is_absolute()) {
            return Err(anyhow!(format!(
                "project file path can not be relative {}",
                file.to_string_lossy()
            )));
        }
        if let None = file.file_name() {
            return Err(anyhow!(format!("project file has no name")));
        }
        self.file = file.clone();
        Ok(())
    }
}

impl ProjectCfg for GlobalProjectCfg {
    fn path(&self) -> PathBuf {
        self.file.to_owned()
    }
}

impl SetupsCfg for GlobalProjectCfg {
    type Setup = GlobalProjectSetupCfg;

    fn get_setups(&self) -> Rc<RefCell<Vec<Rc<RefCell<Self::Setup>>>>> {
        Rc::clone(&self.setups)
    }
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    use crate::cfg::global::project::GlobalProjectCfg;
    use crate::cfg::global::setup::GlobalProjectSetupCfg;
    use crate::cfg::{EnvPathCfg, SetupsCfg};

    #[test]
    fn global_update_private_env_dir() {
        let setup_cfg = GlobalProjectSetupCfg::new("setup".into());

        let mut project_cfg = GlobalProjectCfg::new(&"/project".into()).unwrap();
        project_cfg.add_setup(setup_cfg);

        assert!(project_cfg.get_setups().borrow().iter().count().eq(&1));

        {
            let setup_cfg = project_cfg.get_setup("setup".into()).unwrap();
            setup_cfg
                .borrow_mut()
                .set_env_path_op(Some("/private_env".into()));
        }

        let global_project_setup_cfg_1 = project_cfg.get_setup("setup".into()).unwrap();
        assert_eq!(
            global_project_setup_cfg_1.borrow().env_path(),
            PathBuf::from("/private_env")
        );

        project_cfg.remove_by_name_setup("setup".into());
        assert!(project_cfg.get_setup("setup".into()).is_none());
    }
}
