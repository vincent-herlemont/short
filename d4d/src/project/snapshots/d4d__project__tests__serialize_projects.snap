---
source: d4d/src/project/mod.rs
expression: "&s"
---
"---\nprojects: []"
